tests/ui/consts/const-eval/ub-ref-ptr.stderr

```diff
@@ -148,11 +148,11 @@ LL | const DATA_FN_PTR: fn() = unsafe { mem::transmute(&13) };
                HEX_DUMP
            }
 
-error: accessing memory with alignment 1, but alignment 4 is required
+error[E0080]: evaluation of constant value failed
```

Comment:
This does make a future-incompat warning into a hard error. Based on the comment [here](https://github.com/rust-lang/rust/pull/104616#issuecomment-1349592936) though, this seems to be pre-approved by T-lang though. In any case, cc @RalfJung and @oli-obk for awareness

Reply:
Yeah, these future incompat warnings were because we wanted to make moving away from dubious const-eval patterns smoother as part of the Const UB Armistice of https://github.com/rust-lang/rust/issues/99923, so that const UB doesn't _immediately_ turn into const-break-the-build due to compiler changes. If people feel it's been enough time we can switch this off.

Reply:
This particular case just affects `ptr::read()` which should definitely be fine imo. I'll leave the bikeshedding about what do to with the other cases to everyone else :)

Reply:
IIRC, general temperature was that for these UB-in-`const` cases, a single warning stable cycle is probably sufficient, two is definitely sufficient, and that we're within rights to do no warning releases if we wanted to (i.e. warning at all is a good faith best effort to give some time to migrate).

Reply:
Amusingly, the "see issue" is pointing at https://github.com/rust-lang/rust/issues/68585, which is

> Tracking issue for conflicting repr(...) hints future compatibility

Reply:
Oh, and [`ptr::read`](https://doc.rust-lang.org/nightly/std/ptr/fn.read.html) is unstable in `const` anyway (#80377), so should have much less impact.

Whereas `copy_nonoverlapping` is const-stable since 1.63, so it might still need time; I'm not sure.

Reply:
Purely if it was up to me: go for it.

It seems like this will be a hugely beneficial change, crates impacted were still technically "doing it wrong", we're landing this no sooner than 1.70 (so they've had 1.68 and will have 1.69 to fix it), and "const-stable since 1.63" actually means that we should cut it off sooner rather than later due to the "Lindy effect" that bad code patterns have (i.e. the longer a pattern exists, the longer it is expected to continue existing).

Reply:
Yeah this is fine, we can by now probably make that entire lint into a hard error.

What I don't understand immediately is why this PR changes behavior here though...

Reply:
It changes behavior because unaligned `copy_nonoverlapping` is (temporarily) allowed, but unaligned derefs are not. And this changes `ptr::read` from using copy_nonoverlapping to a deref: https://godbolt.org/z/M6f5MrjKo

```
error: accessing memory with alignment 1, but alignment 4 is required
 --> /rustc/8a73f50d875840b8077b8ec080fa41881d7ce40d/library/core/src/intrinsics.rs:2393:9
  |
  = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
  = note: for more information, see issue #68585 <https://github.com/rust-lang/rust/issues/104616>
note: inside `copy_nonoverlapping::<u32>`
 --> /rustc/8a73f50d875840b8077b8ec080fa41881d7ce40d/library/core/src/intrinsics.rs:2393:9
note: inside `COPY_NONOVERLAPPING`
 --> <source>:8:5
  |
8 |     ptr::copy_nonoverlapping(unaligned, ptr::addr_of_mut!(dest), 1);
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  = note: `#[deny(invalid_alignment)]` on by default

error[E0080]: evaluation of constant value failed
  --> <source>:14:5
   |
14 |     *unaligned
   |     ^^^^^^^^^^ accessing memory with alignment 1, but alignment 4 is required
```

Reply:
> It changes behavior because unaligned copy_nonoverlapping is (temporarily) allowed, but unaligned derefs are not

That's the thing, unaligned derefs should also be temporarily allowed...